File: online_taksi_flutter/lib/repositories/driver_home_screen.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:geolocator/geolocator.dart';

import '../core/session_manager.dart';
import '../models/driver_models.dart';
import '../models/ride_models.dart';
import '../network/api_client.dart';
import '../repositories/driver_repository.dart';
import '../screens/login_screen.dart';
import '../socket/socket_manager.dart';

class DriverHomeScreen extends StatefulWidget {
  const DriverHomeScreen({super.key});

  @override
  State<DriverHomeScreen> createState() => _DriverHomeScreenState();
}

class _DriverHomeScreenState extends State<DriverHomeScreen> {
  final sessionManager = SessionManager();

  late final DriverRepository driverRepository;
  late final SocketManager socketManager;

  bool socketConnected = false;
  bool isOnline = false;
  bool isLoadingRides = false;
  bool isSendingLocation = false;

  String locationText = "Konum: -";
  String logText = "Hazır";

  List<AvailableRideItem> availableRides = [];
  RideResponse? activeRide;

  @override
  void initState() {
    super.initState();

    driverRepository = DriverRepository(
      apiClient: ApiClient(sessionManager: sessionManager),
    );

    socketManager = SocketManager(sessionManager: sessionManager);

    socketManager.onConnected = () {
      setState(() {
        socketConnected = true;
        logText = "Socket bağlandı";
      });
    };

    socketManager.onDisconnected = () {
      setState(() {
        socketConnected = false;
        logText = "Socket kapandı";
      });
    };

    socketManager.onError = (error) {
      _showMessage(error);
      setState(() {
        logText = "Socket hata: $error";
      });
    };

    socketManager.onMessage = (message) {
      _handleSocketMessage(message);
    };
  }

  Future<void> _connectSocket() async {
    await socketManager.connect();
  }

  Future<void> _setOnline(bool value) async {
    try {
      final response = await driverRepository.setOnline(value);

      setState(() {
        isOnline = response.isOnline;
        logText = value ? "Online oldun" : "Offline oldun";
      });

      _showMessage(value ? "Online oldun" : "Offline oldun");
    } catch (e) {
      _showMessage(e.toString());
    }
  }

  Future<void> _loadAvailableRides() async {
    setState(() => isLoadingRides = true);

    try {
      final response = await driverRepository.getAvailableRides();

      setState(() {
        availableRides = response.rides;
        logText = "Açık ride listesi güncellendi";
      });
    } catch (e) {
      _showMessage(e.toString());
    } finally {
      if (mounted) {
        setState(() => isLoadingRides = false);
      }
    }
  }

  Future<void> _acceptRide(AvailableRideItem ride) async {
    try {
      final accepted = await driverRepository.acceptRide(ride.id);

      setState(() {
        activeRide = accepted;
        availableRides.removeWhere((item) => item.id == ride.id);
        logText = "Ride kabul edildi. id=${ride.id}";
      });

      _showMessage("Ride kabul edildi");
    } catch (e) {
      _showMessage(e.toString());
    }
  }

  Future<void> _updateStatus(String status, String note) async {
    final ride = activeRide;

    if (ride == null) {
      _showMessage("Aktif ride yok");
      return;
    }

    try {
      final updated = await driverRepository.updateRideStatus(
        rideId: ride.id,
        status: status,
        note: note,
      );

      setState(() {
        activeRide = updated;
        logText = "Status güncellendi: $status";
      });

      _showMessage("Status güncellendi");
    } catch (e) {
      _showMessage(e.toString());
    }
  }

  Future<void> _sendCurrentLocation() async {
    setState(() => isSendingLocation = true);

    try {
      final permissionOk = await _ensureLocationPermission();
      if (!permissionOk) {
        _showMessage("Konum izni gerekli");
        return;
      }

      final position = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high,
      );

      await driverRepository.updateLocation(
        lat: position.latitude,
        lng: position.longitude,
      );

      setState(() {
        locationText = "Konum: ${position.latitude}, ${position.longitude}";
        logText = "Konum gönderildi";
      });
    } catch (e) {
      _showMessage(e.toString());
    } finally {
      if (mounted) {
        setState(() => isSendingLocation = false);
      }
    }
  }

  Future<void> _startLocationStream() async {
    final permissionOk = await _ensureLocationPermission();
    if (!permissionOk) {
      _showMessage("Konum izni gerekli");
      return;
    }

    _showMessage("Canlı konum gönderimi başladı");

    Geolocator.getPositionStream(
      locationSettings: const LocationSettings(
        accuracy: LocationAccuracy.high,
        distanceFilter: 10,
      ),
    ).listen((position) async {
      try {
        await driverRepository.updateLocation(
          lat: position.latitude,
          lng: position.longitude,
        );

        if (!mounted) return;

        setState(() {
          locationText = "Konum: ${position.latitude}, ${position.longitude}";
          logText = "Canlı konum gönderildi";
        });
      } catch (e) {
        if (!mounted) return;
        setState(() {
          logText = "Konum gönderme hatası: $e";
        });
      }
    });
  }

  Future<bool> _ensureLocationPermission() async {
    final serviceEnabled = await Geolocator.isLocationServiceEnabled();
    if (!serviceEnabled) {
      _showMessage("Konum servisi kapalı");
      return false;
    }

    var permission = await Geolocator.checkPermission();

    if (permission == LocationPermission.denied) {
      permission = await Geolocator.requestPermission();
    }

    if (permission == LocationPermission.denied ||
        permission == LocationPermission.deniedForever) {
      return false;
    }

    return true;
  }

  void _handleSocketMessage(String message) {
    setState(() {
      logText = message;
    });

    try {
      final json = jsonDecode(message) as Map<String, dynamic>;
      final event = json["event"]?.toString();

      if (event == "NEW_RIDE_REQUEST") {
        final data = json["data"] as Map<String, dynamic>?;

        if (data == null) return;

        final ride = AvailableRideItem.fromJson(data);

        final alreadyExists = availableRides.any((item) => item.id == ride.id);

        if (!alreadyExists) {
          setState(() {
            availableRides.insert(0, ride);
          });
        }

        _showMessage("Yeni ride geldi");
      }
    } catch (e) {
      setState(() {
        logText = "Socket parse hata: $e";
      });
    }
  }

  Future<void> _logout() async {
    socketManager.disconnect();
    await sessionManager.clear();

    if (!mounted) return;

    Navigator.pushAndRemoveUntil(
      context,
      MaterialPageRoute(builder: (_) => const LoginScreen()),
          (_) => false,
    );
  }

  void _showMessage(String text) {
    if (!mounted) return;

    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text(text)),
    );
  }

  @override
  void dispose() {
    socketManager.disconnect();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: _appBar(),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            _statusSection(),
            const SizedBox(height: 12),
            _connectionButtons(),
            const SizedBox(height: 16),
            _locationButtons(),
            const SizedBox(height: 20),
            _activeRideCard(),
            const SizedBox(height: 20),
            _availableRidesSection(),
          ],
        ),
      ),
    );
  }

  AppBar _appBar() {
    return AppBar(
      title: const Text("Driver Home"),
      actions: [
        IconButton(
          onPressed: _logout,
          icon: const Icon(Icons.logout),
        ),
      ],
    );
  }

  Widget _statusSection() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(socketConnected ? "Socket: Bağlı" : "Socket: Bağlı değil"),
        Text(isOnline ? "Durum: Online" : "Durum: Offline"),
        Text(locationText),
        Text(
          "Log: $logText",
          maxLines: 3,
          overflow: TextOverflow.ellipsis,
        ),
      ],
    );
  }

  Widget _connectionButtons() {
    return Column(
      children: [
        SizedBox(
          width: double.infinity,
          child: FilledButton(
            onPressed: _connectSocket,
            child: const Text("Socket Bağlan"),
          ),
        ),
        Row(
          children: [
            Expanded(
              child: FilledButton(
                onPressed: () => _setOnline(true),
                child: const Text("ONLINE OL"),
              ),
            ),
            const SizedBox(width: 8),
            Expanded(
              child: OutlinedButton(
                onPressed: () => _setOnline(false),
                child: const Text("OFFLINE OL"),
              ),
            ),
          ],
        ),
        SizedBox(
          width: double.infinity,
          child: OutlinedButton(
            onPressed: isLoadingRides ? null : _loadAvailableRides,
            child: isLoadingRides
                ? const CircularProgressIndicator()
                : const Text("Açık Ride'ları Getir"),
          ),
        ),
      ],
    );
  }

  Widget _locationButtons() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const Text(
          "Konum İşlemleri",
          style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
        ),
        Row(
          children: [
            Expanded(
              child: FilledButton(
                onPressed: isSendingLocation ? null : _sendCurrentLocation,
                child: isSendingLocation
                    ? const CircularProgressIndicator()
                    : const Text("Konum Gönder"),
              ),
            ),
            const SizedBox(width: 8),
            Expanded(
              child: OutlinedButton(
                onPressed: _startLocationStream,
                child: const Text("Canlı Başlat"),
              ),
            ),
          ],
        ),
      ],
    );
  }

  Widget _activeRideCard() {
    final ride = activeRide;

    if (ride == null) {
      return const Card(
        child: Padding(
          padding: EdgeInsets.all(16),
          child: Text("Aktif ride yok"),
        ),
      );
    }

    return Card(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            const Text(
              "Aktif Ride",
              style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
            ),
            Text("Ride ID: ${ride.id}"),
            Text("Pickup: ${ride.pickupAddress}"),
            Text("Dropoff: ${ride.dropoffAddress}"),
            Text("Durum: ${ride.status}"),
            const SizedBox(height: 12),
            _statusButton("Yoldayım", "DRIVER_ARRIVING", "Şoför müşteriye doğru yola çıktı."),
            _statusButton("Geldim", "DRIVER_ARRIVED", "Şoför alım noktasına ulaştı."),
            _statusButton("Başlat", "RIDE_STARTED", "Müşteri araca bindi."),
            _statusButton("Bitir", "RIDE_COMPLETED", "Yolculuk tamamlandı."),
          ],
        ),
      ),
    );
  }

  Widget _statusButton(String title, String status, String note) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 8),
      child: SizedBox(
        width: double.infinity,
        child: FilledButton(
          onPressed: () => _updateStatus(status, note),
          child: Text(title),
        ),
      ),
    );
  }

  Widget _availableRidesSection() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const Text(
          "Açık Ride Listesi",
          style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
        ),
        const SizedBox(height: 8),
        if (availableRides.isEmpty)
          const Text("Açık ride yok")
        else
          ...availableRides.map(_rideCard),
      ],
    );
  }

  Widget _rideCard(AvailableRideItem ride) {
    return Card(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text("Ride ID: ${ride.id}", style: const TextStyle(fontWeight: FontWeight.bold)),
            Text("Pickup: ${ride.pickupAddress}"),
            Text("Dropoff: ${ride.dropoffAddress}"),
            Text("Tahmini Ücret: ${ride.estimatedFare ?? "-"}"),
            const SizedBox(height: 12),
            SizedBox(
              width: double.infinity,
              child: FilledButton(
                onPressed: () => _acceptRide(ride),
                child: const Text("Kabul Et"),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

File: OnlineTaksiJetpack/app/src/main/java/com/example/onlinetaksijetpack/ui/driver/DriverHomeScreen.kt
package com.example.onlinetaksijetpack.ui.driver


import android.Manifest
import android.widget.Toast
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.Card
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.example.onlinetaksijetpack.data.local.SessionManager
import com.example.onlinetaksijetpack.data.remote.api.ApiClient
import com.example.onlinetaksijetpack.data.repository.DriverRepository
import com.example.onlinetaksijetpack.location.DriverLocationTracker

@Composable
fun DriverHomeScreen() {
    val context = LocalContext.current
    val sessionManager = remember { SessionManager(context) }
    val locationTracker = remember { DriverLocationTracker(context) }

    val viewModel: DriverHomeViewModel = viewModel(
        factory = DriverHomeViewModelFactory(
            DriverRepository(ApiClient.create(context))
        )
    )

    val uiState by viewModel.uiState.collectAsState()

    val permissionLauncher =
        rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
            if (granted) {
                locationTracker.start { lat, lng ->
                    viewModel.sendLocation(lat, lng)
                }
            } else {
                Toast.makeText(context, "Konum izni gerekli", Toast.LENGTH_SHORT).show()
            }
        }

    LaunchedEffect(uiState.message) {
        uiState.message?.let {
            Toast.makeText(context, it, Toast.LENGTH_SHORT).show()
            viewModel.clearMessage()
        }
    }

    DisposableEffect(Unit) {
        onDispose {
            locationTracker.stop()
        }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(20.dp),
        verticalArrangement = Arrangement.Top
    ) {
        Text(
            text = "Driver Home",
            style = MaterialTheme.typography.headlineMedium
        )

        Spacer(modifier = Modifier.height(16.dp))

        Text(text = if (uiState.isOnline) "Durum: Online" else "Durum: Offline")
        Text(text = "Konum: ${uiState.currentLat}, ${uiState.currentLng}")
        Text(text = "Log: ${uiState.lastLog}")

        Spacer(modifier = Modifier.height(12.dp))

        Button(
            onClick = { viewModel.connectSocket(sessionManager.getToken()) },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Socket Bağlan")
        }

        Spacer(modifier = Modifier.height(8.dp))

        Button(
            onClick = { viewModel.setOnline(true) },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("ONLINE OL")
        }

        Spacer(modifier = Modifier.height(8.dp))

        OutlinedButton(
            onClick = { viewModel.setOnline(false) },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("OFFLINE OL")
        }

        Spacer(modifier = Modifier.height(12.dp))

        Button(
            onClick = { viewModel.loadAvailableRides() },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Açık Ride'ları Getir")
        }

        Spacer(modifier = Modifier.height(8.dp))

        OutlinedButton(
            onClick = { viewModel.loadActiveRide() },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Aktif Ride'ı Getir")
        }

        Spacer(modifier = Modifier.height(12.dp))

        Button(
            onClick = {
                permissionLauncher.launch(Manifest.permission.ACCESS_FINE_LOCATION)
            },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Konum Başlat")
        }

        Spacer(modifier = Modifier.height(8.dp))

        OutlinedButton(
            onClick = { locationTracker.stop() },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Konum Durdur")
        }

        Spacer(modifier = Modifier.height(20.dp))
        HorizontalDivider()
        Spacer(modifier = Modifier.height(16.dp))

        Text(
            text = "Aktif Ride",
            style = MaterialTheme.typography.titleMedium
        )

        Spacer(modifier = Modifier.height(8.dp))

        if (uiState.activeRide == null) {
            Text("Aktif ride yok")
        } else {
            Card(modifier = Modifier.fillMaxWidth()) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Text("Ride ID: ${uiState.activeRide!!.id}")
                    Text("Pickup: ${uiState.activeRide!!.pickup_address}")
                    Text("Dropoff: ${uiState.activeRide!!.dropoff_address}")
                    Text("Durum: ${uiState.activeRide!!.status}")

                    Spacer(modifier = Modifier.height(12.dp))

                    Button(
                        onClick = {
                            viewModel.updateRideStatus(
                                status = "DRIVER_ARRIVING",
                                note = "Şoför müşteriye doğru yola çıktı."
                            )
                        },
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text("Yoldayım")
                    }

                    Spacer(modifier = Modifier.height(8.dp))

                    Button(
                        onClick = {
                            viewModel.updateRideStatus(
                                status = "DRIVER_ARRIVED",
                                note = "Şoför alım noktasına ulaştı."
                            )
                        },
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text("Geldim")
                    }

                    Spacer(modifier = Modifier.height(8.dp))

                    Button(
                        onClick = {
                            viewModel.updateRideStatus(
                                status = "RIDE_STARTED",
                                note = "Müşteri araca bindi."
                            )
                        },
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text("Başlat")
                    }

                    Spacer(modifier = Modifier.height(8.dp))

                    Button(
                        onClick = {
                            viewModel.updateRideStatus(
                                status = "RIDE_COMPLETED",
                                note = "Yolculuk tamamlandı."
                            )
                        },
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text("Bitir")
                    }
                }
            }
        }

        Spacer(modifier = Modifier.height(20.dp))
        HorizontalDivider()
        Spacer(modifier = Modifier.height(16.dp))

        Text(
            text = "Açık Ride Listesi",
            style = MaterialTheme.typography.titleMedium
        )

        Spacer(modifier = Modifier.height(8.dp))

        if (uiState.isLoadingAvailableRides) {
            CircularProgressIndicator()
        } else {
            uiState.availableRides.forEach { ride ->
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(bottom = 12.dp)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text("Ride ID: ${ride.id}")
                        Text("Pickup: ${ride.pickup_address}")
                        Text("Dropoff: ${ride.dropoff_address}")
                        Text("Tahmini Ücret: ${ride.estimated_fare ?: "-"}")

                        Spacer(modifier = Modifier.height(12.dp))

                        Button(
                            onClick = { viewModel.acceptRide(ride.id) },
                            modifier = Modifier.fillMaxWidth(),
                            enabled = !uiState.isAcceptingRide
                        ) {
                            Text("Kabul Et")
                        }
                    }
                }
            }
        }

        Spacer(modifier = Modifier.height(24.dp))
    }
}